diff --git a/scripts/modis_download.py b/scripts/modis_download.py
index b76e16c..b843da8 100755
--- a/scripts/modis_download.py
+++ b/scripts/modis_download.py
@@ -48,10 +48,10 @@ def main():
     parser.add_option("-I", "--input", dest="input", action="store_true",
                       help="insert user and password from standard input")
     # password
-    parser.add_option("-P", "--password", dest="password", required=True,
+    parser.add_option("-P", "--password", dest="password",
                       help="password to connect to the server")
     # username
-    parser.add_option("-U", "--username", dest="user", required=True,
+    parser.add_option("-U", "--username", dest="user",
                       help="username to connect to the server")
     # tiles
     parser.add_option("-t", "--tiles", dest="tiles", default=None,
